src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionContext.FeatureCollection.cs

```diff
@@ -1,35 +1,45 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System;
 using System.Net.Sockets;
 using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
 using Microsoft.AspNetCore.Connections;
 using Microsoft.AspNetCore.Connections.Features;
 using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Features;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Internal
 {
-    internal sealed partial class QuicConnectionContext : IProtocolErrorCodeFeature, ITlsConnectionFeature
+    internal sealed partial class QuicConnectionContext : IProtocolErrorCodeFeature, ITlsConnectionFeature, ITlsApplicationProtocolFeature
     {
+
+        private X509Certificate2? _clientCert;
+
         public long Error { get; set; }
 
         // Support accessing client certificate
         // https://github.com/dotnet/aspnetcore/issues/34756
         public X509Certificate2? ClientCertificate
         {
-            get => throw new NotSupportedException();
-            set => throw new NotSupportedException();
+            get { return _clientCert ??= (X509Certificate2?)_connection.RemoteCertificate; }
+            set { _clientCert = !(_clientCert is null) ? value : _clientCert; }
```

Comment:
What's going on here? Why not:
```suggestion
            set { _clientCert = value; }
```

Reply:
Setting it only if it hasn't already been initialized. It seems weird to me that we'd allow setting it at all since we can't change the underlying connection, shouldn't it always be `_connection.RemoteCertificate`?

Reply:
`_clientCert` should be considered application state, not transport state. E.g. if the app aquires the client cert through some other mechanism it would still store it in `_clientCert`.